workflow/controller/operator_test.go

```diff
@@ -375,6 +376,22 @@ func TestGlobalParams(t *testing.T) {
 	assert.Contains(t, woc.globalParams, "workflow.labels.workflows.argoproj.io/phase")
 }
 
+func TestTemplateName(t *testing.T) {
+	for input, expected := range map[string]string{
+		`template-name-1234567890`:       `template-name`, // Cronworkflows run on a cron
+		`template-name1234567890`:        `template-name`, // Cronworkflow without trailing -
+		`foobar-fs4hy`:                   `foobar`,        // Standard template submission
+		`foobar-foobar-ashda`:            `foobar-foobar`,
+		`foobarfs4hy`:                    `foobar`, // Template without trailing -
+		`foobar`:                         `foobar`, // We don't replace the end of this
+		`foo-bar`:                        `foo-bar`,
+		`foobar-foobar-0987654321-4ab1a`: `foobar-foobar`, // Cronworkflow, resubmitted
+		`foobar-foobar`:                  `foobar-f`,      // Can't really prevent this kind of thing, not desired though
+	} {
+		assert.Equal(t, expected, getTemplateNameFromName(input))
+	}
+}
```

Comment:
So, it looks like if there's a Workflow whose actual name is for example `template-name-1234567890`, we will assume that the origin is `template-name`, right?

Reply:
Yes, that's the point of these tests. Show how originName will behave given a specific name being fed to it as the workflow name.

Reply:
Oh, sorry, let me clarify my question. What I meant was, say there's a Workflow that wasn't created from a WorkflowTemplate or a CronWorkflow, and it just happens to be named in a way in which it has the "-" followed by a certain number of letters or numbers. In that case, the actual origin should be the entire Workflow name, right?

Reply:
It should, but it won't be. We're reverse engineering the name. The same might happen for a workflow submitted with a `generateName:` or even just a name that matches. I'd prefer to fix this with some "buyer beware" stickers in the documentation - if you want to fool this you're welcome to but you won't get what you want. We could get the controller generated workflows to add annotations of their original name so that we can then re-injest that as the correct name, but that is a fair amount of work to avoid users deliberately being silly when they can just read the instructions instead and not fool the regex. And it still won't work in all cases where workflows are being injected by other tooling including kubectl and argo-events.

Reply:
The result of a Workflow submitted with a `generateName` might actually be pretty good - if your `generateName` is `foobar-` I guess the `origin` would be `foobar`, right?

Reply:
@terrytangyuan what do you think? Is there any precedence for template variables that come with a "buyer beware" sticker (i.e. don't use in certain cases or you might get the wrong result)?

Reply:
Just thinking about the alternative - I guess somebody would need to include their own sprig function to perform the same transformation you did, right? I recognize that would make the Workflow more complex.

Reply:
> The result of a Workflow submitted with a `generateName` might actually be pretty good - if your `generateName` is `foobar-` I guess the `origin` would be `foobar`, right?

Yes. But it's impossible to distinguish between `name: foobar-123467890` and `generateName: foobar-`.

Reply:
> Just thinking about the alternative - I guess somebody would need to include their own sprig function to perform the same transformation you did, right? I recognize that would make the Workflow more complex.

They can already do this, and are... see the issue #6772 this is trying to assist with. We don't need to support overriding the regex in this facility as users can and are already implementing that.

The reliable answer is to use the alternative I've provided in this PR, the entrypoint name instead.

Reply:
I like the entrypoint functionality. Are you thinking of maybe just including that in this PR then?

Reply:
I'd prefer to include both.
`mainEntrypoint` is the 'right answer', but it requires work and requires that users write all their workflows with a differing entrypoint, which is not initially likely, as probably they all have them called `main` or at least haven't given them good, unique names. Adding controls to ensure `mainEntrypoint` is unique for all your workflows ever is problematic.
Because of this, the alternative `originName` is a quick fix for people who want a nice distinguishing name with minimal effort and has been shown to work in the field effectively. We can then determine whether we should put more effort into that solution when issues arise, but also point at `mainEntrypoint` in the docs. Perhaps we can implement workflow automatically generated annotations from common sources if it proved a popular but flawed feature.

Let me do the documentation that's needed from Terry's first comment, and that might help.

Reply:
Okay, thanks. I'll leave it to @terrytangyuan to decide if we should include a variable that should only be used in certain circumstances and won't work if used in the wrong circumstance, if it provides value to people. (I'm not sure if we've come across that type of thing before)